tc/core/polyhedral/cuda/mapped_scop.cc

```diff
@@ -316,62 +312,49 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
   return st->ancestor(root, 2);
 }
 
-size_t MappedScop::mapToThreads(detail::ScheduleTree* band, size_t nInner) {
+size_t MappedScop::mapToThreads(detail::ScheduleTree* band) {
   using namespace tc::polyhedral::detail;
 
-  if (nInner >= numThreads.view.size()) {
-    return nInner;
-  }
-
   bool mapReduction = reductionBandUpdates_.count(band) == 1;
   if (mapReduction) {
     // A reduction is assumed to get mapped to threadIdx.x
-    if (nInner != 0) {
-      reductionBandUpdates_.erase(band);
-      return nInner;
-    }
     CHECK(reductionBandUpdates_.at(band).separated);
     threadIdxXScheduleDepthState.emplace_back(std::make_pair(
         activeDomainPoints(schedule(), band),
         band->scheduleDepth(schedule()) + 0));
     auto reductionDim = reductionBandUpdates_.at(band).reductionDim;
     band = map(band, reductionDim, mapping::ThreadId::x());
-    markUnroll(scop_->scheduleRoot(), band, unroll);
-    return 1;
   }
+
   auto bandNode = band->elemAs<ScheduleTreeElemBand>();
-  // If any inner node was mapped to threads and
-  // the current node has a non-coincident member,
-  // then synchronization needs to be introduced.
-  // This also implies that the mapping needs to be completed first.
-  if (anyNonCoincidentMember(bandNode) && nInner > 0) {
-    // Since some thread identifiers were mapped already (nInner > 0),
-    // the band should have descendants.  Double check.
-    CHECK_EQ(band->numChildren(), 1);
-    mapRemaining<mapping::ThreadId>(
-        band->child({0}), nInner, numThreads.view.size());
-    scop_->insertSyncAfter(band->child({0}));
-    return numThreads.view.size();
-  }
   // With current isl scheduler, if coincident dimensions exist in a band,
   // they are outermost.
   // If a band has more than 3 coincident dimensions, this will choose
   // outermost, but we may also want innermost.
   auto nOuterCoincident = bandNode->nOuterCoincident();
+  int firstThreadIndex = mapReduction ? 1 : 0;
   if (!bandNode->permutable_ || nOuterCoincident < 1) {
-    return nInner;
+    return firstThreadIndex;
```

Comment:
I think this deserves a bit more explanation.
It is only correct because `MappedScop::detectReductions` has already done
the required checks.  Actually, it hasn't completely.
`MappedScop::detectReductions` doesn't have the permutable_` check.
That should be fixed first.